bridges/snowbridge/pallets/ethereum-client/src/lib.rs

```diff
@@ -497,10 +497,35 @@ pub mod pallet {
 			let header_root: H256 =
 				header.hash_tree_root().map_err(|_| Error::<T>::HeaderHashTreeRootFailed)?;
 
-			<FinalizedBeaconStateBuffer<T>>::insert(
-				header_root,
-				CompactBeaconState { slot: header.slot, block_roots_root },
-			);
+			let index = FinalizedBeaconStateIndex::<T>::get();
+			// If the ringbuffer is at the first slot, then the previous head is the last slot
+			// because the ringbuffer wraps around
+			let second_last_index =
+				if index == 0 { MaxFinalizedHeadersToKeep::<T>::get() - 1 } else { index - 1 };
+
+			let prev_head_header_root = FinalizedBeaconStateMapping::<T>::get(second_last_index);
+			let prev_head_finalized_state = FinalizedBeaconState::<T>::get(prev_head_header_root)
+				.unwrap_or(CompactBeaconState::default());
+
+			// If the last finalized header and the one provided to be added to the store
+			// is less than a sync committee apart, overwrite the last finalized header since
+			// we only need one finalized update per SLOTS_PER_HISTORICAL_ROOT. If the slots
+			// are SLOTS_PER_HISTORICAL_ROOT apart, store the update in a new index in the
+			// ringbuffer.
```

Comment:
I would suggest to move/refactor these logic to the impl of ringbuffer, a custom `RingBufferMapImplWithConditionalOverWrite` maybe. With a type param `CheckOverWriteTrait` as follows.

```
pub trait CheckOverWriteTrait<Value> {
	fn can_over_write(value: Value, prev_value: Value) -> bool;
}
```